Use latest ubuntu public ami

peterdenham opened the pull request:
This change makes a decision to always use the latest public Ubuntu AMI for launching an ec2 instance.

Review: @jasonswett 

Reference: https://aws.amazon.com/blogs/compute/query-for-the-latest-amazon-linux-ami-ids-using-aws-systems-manager-parameter-store/

peterdenham committed:
use latest amazon ami

jasonswett commented:
Yay, first pull request! Thanks for doing this. I'll take a look as soon as I get the chance, hopefully later today.

One thing I notice right now: it looks like this AMI is for Amazon Linux which is (I understand) kind of a flavor of CentOS, but I wanted the default AMI to be an Ubuntu one. I don't have a conceptual problem with changing the way the AMI is determined but I would prefer that it be Ubuntu.

peterdenham commented:
Thanks and agreed, Ubuntu sounds like a reasonable default here.

jasonswett commented on lib/exosuit/instance.rb:
```diff
@@ -44,5 +44,9 @@ def self.all
       all.select { |i| i.state.name == 'running' }
     end
+
+    def self.latest_ami
```
I vote we instead call this `latest_ami_image_id` since the thing it returns is not an AMI but the image id of an AMI.

jasonswett commented on lib/exosuit/instance.rb:
```diff
@@ -44,5 +44,9 @@ def self.all
     end
+
+    def self.latest_ami
+      @ami = Exosuit.ssm.get_parameters({ names: [IMAGE_NAME] }).parameters[0].value
```
If this is the one and only place we use SSM, how about we instead do this?

```ruby
Aws::SSM::Client.new(profile: profile_name)
  .get_parameters(names: [IMAGE_NAME])
  .parameters[0]
  .value
```

Then we don't have to have the `Exosuit#ssm` method which may never be used anywhere else.

jasonswett reviewed (changes requested):
Looks pretty good! If we can just change the AMI type from Amazon Linux to Ubuntu and make the other code changes, I'll be ready to merge!

peterdenham committed:
PR Feedback

peterdenham commented:
Addressed your comments, but I need to revisit my approach on how to determine the latest ami for Ubuntu. They currently only support listing Amazon Linux and Windows ami versions in parameter store.

peterdenham committed:
Find latest ubuntu bionic image

peterdenham commented:
Updated logic to retrieve the latest `ubuntu-bionic-18.04-amd64` image. While this wildcard name matching approach could be error prone, I checked the image release history and every release since the first bionic release on `2018-04-26`, has been consistently released with that naming structure.

Another open question is if you have a preference for the x86 or arm architecture - this assumes x86.

Ready for another look when you get a chance! @jasonswett

peterdenham commented:
I debated whether it was worthwhile using this approach or just a hard coding a known "good" ami. Ubuntu seems to release a new image every 1-3 weeks, so it feels like this is a good decision to get new projects launched on the latest image without having to manually update the source.

peterdenham committed:
Remove SSM dependency

jasonswett commented on lib/exosuit/instance.rb:
```diff
@@ -44,5 +44,14 @@ def self.all
     end
+
+    def self.latest_ubuntu_ami
+      ubuntu_image ||= Exosuit.ec2.images(
```
I imagine your goal with `ubuntu_image ||=` is to cache the value of the variable. I don't believe a local variable will achieve that effect, I believe you need to use an instance variable. I'd also vote that we use the same name as the method name, e.g. `@latest_ubuntu_ami ||=`, which is kind of a convention I use everywhere.

jasonswett commented on lib/exosuit/instance.rb:
```diff
@@ -44,5 +44,14 @@ def self.all
+              executable_users: ["all"],
+              filters: [ { name: "name",values: [IMAGE_NAME]}]
+          }
+      ).sort_by {|i| i.creation_date}.last
```
Just a few style items (I need to get Rubocop or something on here to automate style stuff):

- Let's use single quotes in places where double quotes aren't needed
- Let's make the spacing more consistent - `[{ name: 'name', values: [IMAGE_NAME] }]`
- I believe you can actually do `sort_by(&:creation_date).last` instead of `sort_by {|i| i.creation_date}.last`

jasonswett reviewed (changes requested):
Look really good, and after a few super minor changes I believe we're all set! Thanks again for contributing. Looking forward to merging this first PR.

peterdenham committed:
PR Feedback

jasonswett commented:
Perfect! Merged!
